src/desktop/lib/psi/buildServerApp.js

```diff
@@ -0,0 +1,87 @@
+import React, { Fragment } from 'react'
+import ReactDOMServer from 'react-dom/server'
+import createRender from 'found/lib/createRender'
+import queryMiddleware from 'farce/lib/queryMiddleware'
+import serialize from 'serialize-javascript'
+import { AppProvider } from './AppProvider'
+import { Resolver } from 'found-relay'
+import { getFarceResult } from 'found/lib/server'
+import { getLoadableState } from 'loadable-components/server'
+
+import { createRelayEnvironment } from './relayEnvironment'
+
+export function buildServerApp(routeConfig, url = '/') {
+  return new Promise(async (resolve, reject) => {
+    try {
+      if (!process.env.SSR_ENABLED) {
+        return resolve({
+          RelayApp: x => x,
+          relayData: {},
+          loadableScripts: {},
+        })
+      }
+
+      const environment = createRelayEnvironment()
+      const historyMiddlewares = [queryMiddleware]
+      const resolver = new Resolver(environment)
+      const render = createRender({})
+
+      const { redirect, status, element } = await getFarceResult({
+        url,
+        historyMiddlewares,
+        routeConfig,
+        resolver,
+        render,
+      })
+
+      // TODO: Deal with redirect one level up?
+      if (redirect) {
+        return resolve({ redirect, status })
+      }
+
+      const App = () => (
+        <AppProvider
+          provide={{
+            environment,
+            resolver,
+            routeConfig,
+          }}
+        >
+          {element}
+        </AppProvider>
+      )
+
+      // Kick off relay requests
+      ReactDOMServer.renderToString(App)
```

Comment:
As noted on another PR (not sure anymore which one), it would be great if we could ensure we don’t actually render any views at all here, e.g. a loading view that really only needs to show when performing requests on the client.

Reply:
Wait, where’s the second `ReactDOMServer.renderToString(App)` call that was needed previously?

Reply:
That's moved out to userland -- `const { App } = await buildServerApp(routes)` -- that way in the end they can do what they want with it, as its just a normal react component.

Reply:
Smart 🙆‍♂️